src/typescript-service.ts

```diff
@@ -338,6 +338,847 @@ export default class TypeScriptService {
 		});
 	}
 
+	getDocumentSymbol(params: DocumentSymbolParams): Promise<SymbolInformation[]> {
+		const uri = params.textDocument.uri;
+		return this.ensureFilesForHoverAndDefinition(uri).then(() => {
+			const fileName = util.uri2path(uri);
+			const config = this.projectManager.getConfiguration(uri);
+			return config.prepare(this.connection).then(() => {
+				this.projectManager.syncConfiguration(config, this.connection);
+				const sourceFile = config.program.getSourceFile(fileName);
+				const tree = config.service.getNavigationTree(fileName);
+				const limit = 100000;
+				const result: SymbolInformation[] = [];
+				this.flattenNavigationTreeItem(tree, null, sourceFile, result, limit);
+				return Promise.resolve(result);
+			});
+		});
+	}
+
+	getWorkspaceReference(params: util.WorkspaceReferenceParams): Promise<util.ReferenceInformation[]> {
+		const refInfo: util.ReferenceInformation[] = [];
+		return this.ensureFilesForWorkspaceSymbol().then(() => {
+			return Promise.all(this.projectManager.getConfigurations().map((config) => {
+				return config.prepare(this.connection).then((config) => {
+					this.projectManager.syncConfiguration(config, this.connection);
+					for (let source of config.service.getProgram().getSourceFiles()) {
+						if (source.fileName.indexOf("/node_modules/") !== -1) {
+							continue;
+						}
+						this.walkMostAST(source, (node) => {
+							switch (node.kind) {
+								case ts.SyntaxKind.Identifier: {
+									const fileName = source.fileName.startsWith("/") ? source.fileName.substr(1) : source.fileName;
+									const id = node as ts.Identifier;
+									const defs = config.service.getDefinitionAtPosition(fileName, node.pos + 1);
+
+									if (defs && defs.length > 0) {
+										const def = defs[0];
+										const start = ts.getLineAndCharacterOfPosition(source, node.pos);
+										const end = ts.getLineAndCharacterOfPosition(source, node.end);
+										const ref = {
+											location: {
+												uri: util.path2uri('', source.fileName),
+												range: {
+													start: start,
+													end: end,
+												},
+											},
+											name: def.name,
+											containerName: def.containerName,
+											uri: util.path2uri('', def.fileName),
+										};
+										refInfo.push(ref);
+									}
+									break;
+								}
+								case ts.SyntaxKind.StringLiteral: {
+									// TODO
+									break;
+								}
+							}
+						});
+					}
+				});
+			}));
+		}).then(() => {
+			return refInfo;
+		});
+	}
+
+	/*
+	 * walkMostAST walks most of the AST (the part that matters for gathering all references)
+	 */
+	walkMostAST(node: ts.Node, visit: (node: ts.Node) => void) {
+		visit(node);
+		const children: ts.Node[] = [];
```

Comment:
We could do it in place, but it makes the code a lot more verbose (need to insert a bunch more calls to `walkMostAST`)